src/controller/ai_controller.c

```diff
@@ -93,9 +108,176 @@ int char_to_act(int ch, int direction) {
     return ACT_STOP;
 }
 
+/** 
+ * \brief Roll chance for pilot preference.
+ *
+ * \param pref_val The value of the pilot preference (-400 to 400)
+ *
+ * \return A boolean indicating whether the preference is confirmed.
+ */
+bool roll_pref(int pref_val) {
+    int rand_roll = rand_int(800);
+    int pref_thresh = pref_val + 400;
+    return rand_roll <= pref_thresh;
+}
+
+/** 
+ * \brief Determine whether the AI is smart enough to usually go ahead with an action.
+ *
+ * \param a The AI instance.
+ *
+ * \return A boolean indicating whether the AI is smart enough.
+ */
+bool smart_usually(ai *a) {
+    int rand_roll = rand_int(32);
+    int diff_factor = a->difficulty * a->difficulty; // 1 - 49
+    return rand_roll <= diff_factor;
+}
+
+/** 
+ * \brief Determine whether the AI is dumb enough to usually go ahead with an action.
+ *
+ * \param a The AI instance.
+ *
+ * \return A boolean indicating whether the AI is dumb enough.
+ */
+bool dumb_usually(ai *a) {
+    return !smart_usually(a);
+}
+
+/** 
+ * \brief Determine whether the AI is smart enough to sometimes go ahead with an action.
+ *
+ * \param a The AI instance.
+ *
+ * \return A boolean indicating whether the AI is smart enough.
+ */
+bool smart_sometimes(ai *a) {
```

Comment:
In functions, please use const keyword when args are not changed.

Reply:
Addressed in comment: https://github.com/omf2097/openomf/pull/387/commits/3e81cf97b840c9abbbdd3afc2f7b6bd0dba2dff7